feat(dyn): use smarter printing of floats

Alizter opened the pull request:
This algorithm is from janestreet/sexplib0.

- [x] Fixes #12980.
- [x] Checked round trip
- [x] changelog

cc @mbarbin

Alizter committed:
feat(dyn): use smarter printing of floats

This algorithm is from janestreet/sexplib0.

Signed-off-by: Ali Caglayan <alizter@gmail.com>

mbarbin commented:
@Alizter I ended up vendoring dyn in a project so it's easy to import diffs directly and see the result in existing tests. See [PR](https://github.com/mbarbin/super-master-mind/pull/36) which is a mirror of your change here.

We can see in the diff in the test that the floats that are integer are no longer printed with a trailing dot. I would not have expecting this (this probably makes sense for sexp but maybe not for dyn where it's better if the expressions look like OCaml litteral). I'll have another look in the next couple days. Feel free to send me PR for that, or experiment with other changes. Thanks!

Alizter commented:
@mbarbin 

> I had a look at the PR, you probably know this but just for completeness, I have memories that in sexplib0 the formatting function used is [caml_format_float](https://github.com/janestreet/sexplib0/blob/0de081f12e5db8307a4ecad197629373426c8688/src/sexp_conv.ml#L54) because it ended up being faster than `Printf.sprintf` - but maybe you prefer not depending on an external, or the gain is not significant.

I didn't bother using that, but I think its fine to avoid the overhead. I've started using it now.

> We can see in the diff in the test that the floats that are integer are no longer printed with a trailing dot. I would not have expecting this (this probably makes sense for sexp but maybe not for dyn where it's better if the expressions look like OCaml litteral). I'll have another look in the next couple days. Feel free to send me PR for that, or experiment with other changes. Thanks!

I've changed some things around from my last push. I've used `g` instead of `G` like sexplib0 does and added some code the check integer formats. I'm sure the code can be optimised a bit more here, so I am open to any ideas.

I've also added some tests so that we can avoiding breaking this round-trip behaviour in dyn in the future.

mbarbin reviewed (approved):
Hi @Alizter the diff looks good to me and I confirm this fixed things on my end. Thanks for initiating a test lib for dyn too, it'll be easy to extend as needed in the future.
